Add customer facing prices for customer multi-currency

luizreis opened the pull request:
Fixes #1769 

## Changes proposed in this Pull Request

This PR converts frontend prices to the selected currency, maintaining compatibility with WooCommerce's native features. Compatibility with other extensions will be addressed as part of another issue. But this work *should* also be compatible with WooCommerce Subscriptions, barring a few edge cases like Subscriptions-specific coupons.

The way prices are converted and formatted is by using some WooCommerce's hooks. Two classes were created to handle them: [Frontend_Prices](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-5e2b0c63cb5e2acb8f645e5505db93b758eb74c36703584c42367c726bb2ed0bR15) and [Frontend_Currencies](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-c67278bce0bbc54e546375c4d665efd19475a58c00f2586e6cc0f01cb981064aR15). As their names might indicate, they are only loaded in the frontend, by using the [same check as WC](https://github.com/woocommerce/woocommerce/blob/6667233eb11b3d925f5df502e4090742bf1b0c0f/includes/class-woocommerce.php#L359). Using that loading check means that, for now, the multi-currency prices won't work for the checkout blocks. Since we're only building support for blocks in the future, I think it'd be best to address this as part of another issue. Is that okay?

The Frontend_Prices class is responsible for the price conversion for products, shipping rates, and coupons. The Frontend_Currencies one is responsible for using the right symbol, formatting, separators, and decimal places when rendering currencies, and uses locale data from [woocommerce/i18n/locale-info.php](https://github.com/woocommerce/woocommerce/blob/b19500728b4b292562afb65eb3a0c0f50d5859de/i18n/locale-info.php). The data in the locale file can be filtered using [`wcpay_multi_currency_<currency_code>_format`](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-c67278bce0bbc54e546375c4d665efd19475a58c00f2586e6cc0f01cb981064aR126).

I made a few assumptions when calculating and adjusting the prices, such as we'd always want to round up the converted amount, we wouldn't want to apply charm pricing to coupons, and we wouldn't want to ceil and apply charm pricing to taxes. Do they seem right?

The price calculation and adjustment can be found [here](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-b17bf67925ece08a12ad748cd870705e5d668d3fa19214040f4841daaf712cdcR263-R326). When adjusting the price, we always round up. However, PHP doesn't provide a `ceil` method that accepts the number of decimal places and I had to write [`ceil_price`](https://github.com/Automattic/woocommerce-payments/pull/1860/files#diff-b17bf67925ece08a12ad748cd870705e5d668d3fa19214040f4841daaf712cdcR323). Is that solution robust and reliable enough for our use case?

A few settings are being mocked currently, and can be changed by using filters:
- `wcpay_multi_currency_round_precision`: The number of decimal places to use when rounding, can be negative.
- `wcpay_multi_currency_charm_pricing`: The charm pricing modifier that is added when adjusting the price.
- `wcpay_multi_currency_apply_charm_only_to_products`: Setting this to false indicates that shipping rates should have a charm modifier as well.

A few of the tests introduced in this PR rely on the mocked currencies setup and will have to be modified when we start fetching currencies for real.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

You can select a different currency by using the `?currency=<code>` URL param when visiting a frontend page.

To be able to test different currency formatting and zero-currencies, I've modified `get_mock_currencies` to include:

```
return [
	[ 'USD', '1.00' ],
	[ 'CHF', '0.89873' ],
	[ 'HUF', '286.33421' ], // Zero decimal currency.
	[ 'KRW', '1127.7053' ], // Zero decimal currency.
	...
```

We need to make sure that the select currency is working with the following WooCommerce's features:

- Price formatting and conversion
    - Load up the Shop page with the default store currency and check that all prices look right
    - Select a different currency and ensure that the prices have been converted and formatted accordingly:
        - `USD` should look like `$x'xxx.xx`
        - `CHF` should look like `CHF x,xxx.xx`
        - `HUF` should look like `x xxx Ft`
        - `KRW` should look like `xx,xxx₩`
    - Make sure that discounted products and variation price ranges are rendered correctly
- Simple products
    - Make sure their price is converted in the Shop, Product, Cart, and Checkout pages
    - Purchase the product and check that the order has the right currency and price
    - Navigate to the transactions page and ensure that the currency used was the selected one
- Variable products
    - Make sure their price is converted in the Shop, Product, Cart, and Checkout pages
    - Modify the selected currency and ensure the price range was updated. It is cached by WC by default but should be refreshed when the exchange rate changes
    - Purchase the product and check that the order has the right currency and price
    - Navigate to the transactions page and ensure that the currency used was the selected one
- Coupons
    - Enable coupon codes in the General WooCommerce settings
    - Create three coupons, one for each type
    - Ensure that the fixed cart discount and fixed product discount amounts are converted when using them
    - Ensure that the percentage discount works correctly, without having its amount converted
    - Add a minimum and maximum spend usage restriction to a coupon, and check that those values are converted and matched against the cart subtotal
- Shipping
    - Add a free and a flat rate shipping method to your shipping zone
    - Make sure that they work correctly on the checkout page
        - If `apply_charm_only_to_products` is enabled, the shipping rate should only be ceiled after conversion
        - If it's not, the shipping rate should also have the charm pricing added to it
- Taxes
    - Enable taxes in the General WooCommerce settings and create a standard tax rate
    - Ensure that taxes are correctly applied and converted when using both "Prices inclusive of tax" and "Prices exclusive of tax"
    - Ensure that the shipping methods' taxes are also converted properly under both configurations


-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

luizreis commented on includes/multi-currency/class-frontend-currencies.php:
```diff
@@ -0,0 +1,155 @@
+		$this->load_locale_data();
+
+		// Currency hooks.
+		add_filter( 'woocommerce_currency', [ $this, 'get_current_currency_code' ] );
+		add_filter( 'wc_get_price_decimals', [ $this, 'get_current_currency_decimals' ] );
+		add_filter( 'wc_get_price_decimal_separator', [ $this, 'get_current_currency_decimal_separator' ] );
+		add_filter( 'wc_get_price_thousand_separator', [ $this, 'get_current_currency_thousand_separator' ] );
+		add_filter( 'woocommerce_price_format', [ $this, 'get_current_currency_format' ] );
```
Should we register these hooks with a higher priority to ensure they're executed after the default priority?

dmallory42 replied:
That makes sense to me even if just to mitigate the risk of weird edge-case issues.

luizreis replied:
👍  Added in https://github.com/Automattic/woocommerce-payments/pull/1860/commits/950f6095631fdd12d953cc7d09b7353a042ccb1d

luizreis commented on includes/multi-currency/class-frontend-prices.php:
```diff
@@ -0,0 +1,156 @@
+	public function __construct( Multi_Currency $multi_currency ) {
+		$this->multi_currency = $multi_currency;
+
+		// Simple product price hooks.
+		add_filter( 'woocommerce_product_get_price', [ $this, 'get_product_price' ] );
+		add_filter( 'woocommerce_product_get_regular_price', [ $this, 'get_product_price' ] );
+		add_filter( 'woocommerce_product_get_sale_price', [ $this, 'get_product_price' ] );
+
+		// Variation price hooks.
+		add_filter( 'woocommerce_product_variation_get_price', [ $this, 'get_product_price' ] );
+		add_filter( 'woocommerce_product_variation_get_regular_price', [ $this, 'get_product_price' ] );
+		add_filter( 'woocommerce_product_variation_get_sale_price', [ $this, 'get_product_price' ] );
+
+		// Variation price range hooks.
+		add_filter( 'woocommerce_variation_prices', [ $this, 'get_variation_price_range' ] );
+		add_filter( 'woocommerce_get_variation_prices_hash', [ $this, 'add_exchange_rate_to_variation_prices_hash' ] );
+
+		// Shipping methods hooks.
+		add_filter( 'woocommerce_package_rates', [ $this, 'get_shipping_rates_prices' ] );
+
+		// Coupon hooks.
+		add_filter( 'woocommerce_coupon_get_amount', [ $this, 'get_coupon_amount' ], 20, 2 );
+		add_filter( 'woocommerce_coupon_get_minimum_amount', [ $this, 'get_coupon_min_max_amount' ] );
+		add_filter( 'woocommerce_coupon_get_maximum_amount', [ $this, 'get_coupon_min_max_amount' ] );
```
Should we register these hooks with a higher priority to ensure they're executed after the default priority?

dmallory42 replied:
As with the other hooks, I think this sounds sensible.

luizreis replied:
Added in https://github.com/Automattic/woocommerce-payments/pull/1860/commits/950f6095631fdd12d953cc7d09b7353a042ccb1d

ismaeldcom commented:
> Using that loading check means that, for now, the multi-currency prices won't work for the checkout blocks. Since we're only building support for blocks in the future, I think it'd be best to address this as part of another issue. Is that okay?

I agree here, best to focus on legacy implementation for now.

> I made a few assumptions when calculating and adjusting the prices, such as we'd always want to round up the converted amount, we wouldn't want to apply charm pricing to coupons, and we wouldn't want to ceil and apply charm pricing to taxes. Do they seem right?

I will say yes to almost all these questions. But I don't know if some customer would like to have more precision. For example, I made an order for an item of `₩2,030,873.90` but paid for `₩2,030,874.00`, the final user saw the rounded price. Should we offer the option to round prices? IMO is not necessary, but I just want to think a bit on other customer needs.

Followed testing instructions, and almost everything went well, just found this 2 little concerns:

- Price formatting and conversion
  - `CHF` should look like `CHF x,xxx.xx`. I see it like `CHF x'xxx.xx` instead. I think you made a mistake writing it here, because in `locale-info.php` the thousand separator is `'` too.
- Shipping. Free Shipping `Minimum order amount` is not working right. Looks like is checking the actual number and not the converted currency. Example:
    - minimum set to $100
    - Cart $110 => shows free shipping
    - Switch to CHF 98.68 => missing free shipping
    - Switch to HUF 31 497 => shows free shipping

Still need to check all the code, but here is a first look 🙂

dmallory42 commented:
> Since we're only building support for blocks in the future, I think it'd be best to address this as part of another issue. Is that okay?

Yep, I also agree with this approach. Do you think we should create a follow-up issue to this just to make sure we keep it in mind? :) 

> I made a few assumptions when calculating and adjusting the prices, such as we'd always want to round up the converted amount, we wouldn't want to apply charm pricing to coupons, and we wouldn't want to ceil and apply charm pricing to taxes. Do they seem right?

This matches my understanding. We should always round up the converted amount to whatever the selected rounding precision is, because this (and charm pricing) are intended as mechanisms to ensure favourable pricing for the merchant. I also agree about not applying these rules to coupons or taxes.

dmallory42 commented on includes/multi-currency/class-frontend-currencies.php:
```diff
@@ -0,0 +1,155 @@
+	 *
+	 * @var array
+	 */
+	protected $currency_settings = [];
```
This is probably overkill, but what do you think about creating a class to represent this rather than using an array?

luizreis replied:
I also think that at this point it might be overkill. Using the array structure might make it easier if people need to use the `wcpay_multi_currency_<currency>_settings`. I'm happy either way though, so let me know if you'd rather implement a class now 🙂

dmallory42 replied:
That makes sense, I'm happy to keep it as an array as you said I think it makes it simpler for people who may need to use the hooks. :)

luizreis commented:
> I will say yes to almost all these questions. But I don't know if some customer would like to have more precision. For example, I made an order for an item of `₩2,030,873.90` but paid for `₩2,030,874.00`, the final user saw the rounded price. Should we offer the option to round prices? IMO is not necessary, but I just want to think a bit on other customer needs.

I'm not sure I understood this, but if you mean that we should offer merchants a way to decide how to round the prices we plan on having that. For now, since the settings page is in progress, I've added a getter method and a way to control it with the `wcpay_multi_currency_round_precision` filter. This filter defaults to 0 and should return the number of decimals to use when rounding. It can be negative to round to tens and hundreds, similarly to [PHP's round $precision](https://www.php.net/manual/en/function.round.php).

>   * `CHF` should look like `CHF x,xxx.xx`. I see it like `CHF x'xxx.xx` instead. I think you made a mistake writing it here, because in `locale-info.php` the thousand separator is `'` too.

Good catch! I've fixed the testing instructions.

> * Shipping. Free Shipping `Minimum order amount` is not working right. Looks like is checking the actual number and not the converted currency. Example:
>   
>   * minimum set to $100
>   * Cart $110 => shows free shipping
>   * Switch to CHF 98.68 => missing free shipping
>   * Switch to HUF 31 497 => shows free shipping

I retested this bit and it is actually checking the converted currency, but I think I was able to figure out what happened in your testing:

The product price gets ceiled and has 10 cents of charm price discounted, while the min coupon spending is only ceiled. I created a product with the same price as the min coupon amount and when charm pricing is enabled the coupon cannot be applied. If I set `wcpay_multi_currency_charm_pricing` to return `0.0` it works. It didn't happen with HUF because the currency has no decimal places. Does the amount in the error message you see when adding match this scenario?

Screenshots: [Coupon amount](https://d.pr/i/WGN08t) | [Charm pricing enabled](https://d.pr/i/yBp3Io) | [Charm pricing disabled](https://d.pr/i/UKFw3E)

I think we might need to change a bit how we're handling the coupon prices to better account for these cases. Let's discuss this in Slack and I'll update the PR with our decisions.

dmallory42 commented on includes/multi-currency/class-multi-currency.php:
```diff
@@ -72,14 +86,24 @@ private function __construct() {
-		add_action( 'init', [ $this, 'update_selected_currency_by_url' ] );
 		add_action( 'rest_api_init', [ __CLASS__, 'init_rest_api' ] );
+
+		$is_frontend_request = ( ! is_admin() || defined( 'DOING_AJAX' ) ) && ! defined( 'DOING_CRON' ) && ! WC()->is_rest_api_request();
```
This is unrelated to this PR, but I'm surprised there's not a generic WC/WP helper function to tell if it is a frontend request.

luizreis commented:
> Shipping. Free Shipping `Minimum order amount` is not working right. Looks like is checking the actual number and not the converted currency. Example:
> * minimum set to $100
> * Cart $110 => shows free shipping
> * Switch to CHF 98.68 => missing free shipping
> * Switch to HUF 31 497 => shows free shipping

Great catch! 🚀 I've addressed this case in https://github.com/Automattic/woocommerce-payments/pull/1860/commits/6c9d387127069c37fc39626b35108eba90c2499b.

ismaeldcom commented:
>@ismaeldcom - great catch! 🚀 I've addressed this case in 6c9d387.

🤔 Maybe I'm missing something. I tried to test it, but since commit 45cc358 I receive a fatal error about `WC->countries` being null. I think is because the instance is not initiated yet, so made a quick try changing Multi_Currency construct `$this->init();` by `add_action( 'woocommerce_init', [ $this, 'init' ] );` and the fatal error disapear.

But, I can't see any difference, after some debugging the code enters the loop of `register_free_shipping_filters` but `get_free_shipping_min_amount` is never called.

ismaeldcom commented on includes/multi-currency/class-frontend-currencies.php:
```diff
@@ -0,0 +1,155 @@
+	 * @return int The number of decimals.
+	 */
+	public function get_current_currency_decimals() {
+		$currency_code = $this->multi_currency->get_selected_currency()->get_code();
```
Little nitpick/question here. Should we use `get_current_currency_code` here and in the following getters? To avoid reusing the same code `$this->multi_currency->get_selected_currency()->get_code()`.

luizreis replied:
I'm fine either way 😄 

I didn't reuse it at first because that method is mainly intended to be used by the `woocommerce_currency` filter and it could change over time.

But, right now all it does is wrap `$this->multi_currency->get_selected_currency()->get_code()` so we might as well just use it, even if it might cause a bigger refactor in the future in case we need to do something else when running the `woocommerce_currency` filter.

Does that make sense? How do you think we should approach using this method?

luizreis replied:
That said, it would be good to rename these methods to tie them a bit more with the hook they're attached to if we treat them as such. E.g. instead of `get_current_currency_code` we should name it `get_woocommerce_currency` or something along those lines.

ismaeldcom replied:
If we probably change this method in the future, better to leave as it is. just looks like a handy getter even it's used in the hook. And agree to rename the methods attached to hooks.

luizreis replied:
Agreed. I've updated the names in https://github.com/Automattic/woocommerce-payments/pull/1860/commits/f5472150d5de5c08b383ab1ce92e8346784937fe. What do you think of them?

ismaeldcom reviewed (approved):
Everything working good now! 🙂
@luizreis Thanks for fixing this little tricky issues.
Apart from the opened comments, I think it's ready to 🚢 !
